src/DaemonExec.ts

```diff
@@ -0,0 +1,188 @@
+import * as Docker from 'dockerode'
+
+interface ContainerModem {
+	demuxStream(
+		stream: NodeJS.ReadWriteStream,
+		stdoutStream: NodeJS.WritableStream,
+		stderrStream: NodeJS.WritableStream,
+	): void
+}
+
+interface DockerModem {
+	followProgress(
+		stream: NodeJS.ReadableStream,
+		onFinished: (err?: Error) => Promise<void>,
+		onProgress: (event: {
+			status: string
+			progressDetail: { current: number; total: number }
+			progress: string
+			id: string
+		}) => void,
+	): void
+}
+
+export interface VolumeDefinition {
+	fromPath: string
+	toPath: string
+}
+
+/**
+ * Returns information about the images
+ * @param docker - The docker daemon to pull information from
+ * @returns an array of ImageInfo
+ */
+export async function listImages(docker: Docker): Promise<Docker.ImageInfo[]> {
+	return docker.listImages({})
+}
+
+/**
+ * Returns information about the containers
+ * @param docker - The docker daemon to pull information from
+ * @returns an array of ContainerInfo
+ */
+export async function listContainers(
+	docker: Docker,
+): Promise<Docker.ContainerInfo[]> {
+	return docker.listContainers()
+}
+
+/**
+ * Starts the import of an image
+ * @param docker - The docker daemon with which the image will be stored
+ * @param name - The name of the image to pull
+ * @returns a ReadableStream to track the progress of the import
+ */
+export async function importImage(
+	docker: Docker,
+	name: string,
+): Promise<NodeJS.ReadWriteStream> {
+	const stream: NodeJS.ReadWriteStream = await (docker.pull(
+		name,
+	) as Promise<NodeJS.ReadWriteStream>)
+	return stream
+}
+
+/**
+ * Imports an image if it does not exist and then calls callbacks for progress and finish events
+ * @param docker - The docker daemon with which the image will be stored
+ * @param name - The name of the image to pull
+ * @param onFinished - a callback for when the image import is finished
+ * @param onProgress - a callback for when the progress of the image import updates
+ */
+export async function callbackImportImage(
+	docker: Docker,
+	name: string,
+	onFinished: (err?: Error) => Promise<void>,
+	onProgress: (event: {
+		status: string
+		progressDetail: { current: number; total: number }
+		progress: string
+		id: string
+	}) => void,
+): Promise<void> {
+	const images = await listImages(docker)
+
+	if (
+		images.findIndex(
```

Comment:
`Array.some` is better than `findIndex` in this case since you don't really care about the index.